File: core/BalancedBinaryTree/BalancedBinaryTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isBalanced(root: TreeNode) -> bool:
    if not root:
        return False
    res = True
    def dfs(node):
        if not node:
            return 0
        left_deep = dfs(node.left)
        right_deep = dfs(node.right)
        nonlocal res
        if abs(left_deep - right_deep) > 1:
            res = False
        return max(left_deep, right_deep) + 1
    dfs(root)
    return res

File: core/BalancedBinaryTree/BalancedBinaryTree.java
package com.dbc;

public class BalancedBinaryTree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private boolean res = true;
    private int dfs(TreeNode node) {
        if (node == null) return 0;
        int leftDeep = dfs(node.left);
        int rightDeep = dfs(node.right);
        if (Math.abs(leftDeep - rightDeep) > 1) this.res = false;
        return Math.max(leftDeep, rightDeep) + 1;
    }

    public boolean isBalanced(TreeNode root) {
        dfs(root);
        return this.res;
    }
}
